design/014-secretstore-generator-v2.md

```diff
@@ -0,0 +1,376 @@
+- [Design Details](#design-details)
+  - [Test Plan](#test-plan)
+    - [Prerequisite testing updates](#prerequisite-testing-updates)
+    - [Unit tests](#unit-tests)
+    - [Integration tests](#integration-tests)
+    - [e2e tests](#e2e-tests)
+  - [Graduation Criteria](#graduation-criteria)
+  - [Upgrade / Downgrade Strategy](#upgrade--downgrade-strategy)
+  - [Monitoring Requirements](#monitoring-requirements)
+  - [Troubleshooting](#troubleshooting)
+- [Drawbacks](#drawbacks)
+- [Alternatives](#alternatives)
+<!-- /toc -->
+
+## Summary
+
+This proposes a v2 architecture and API for SecretStores and Generators in External Secrets Operator. The primary goals are to:
+
+- Support out-of-tree providers as first-class citizens, allowing independent versioning and distribution.
+- Unify feature sets of SecretStores and Generators (e.g. refresh, gating) under consistent CRDs and controllers.
+- Make referent authentication modes explicit and easier to use.
+- Allow users to install only the providers they need.
+- reduce the dependency footprint of ESO to improve its security posture
+
+There are several limitations in the current (v1) SecretStore and Generator architectures that hinder flexibility and maintainability:
+
+- Different provider versioning is not possible.
+- Out-of-tree providers are not first-class - they're only supported through the webhook provider.
+- Users cannot easily install/uninstall only the desired providers.
+- Referent authentication modes are implicit and hard to learn/use.
+
+## Goals
+
+- Implement new CRDs: SecretStore/v2alpha1, ClusterSecretStore/v2alpha1, Generator/v2alpha1, ClusterGenerator/v2alpha1.
+- Enable ESO to run without in-tree providers; users install providers separately.
+- Provide a provider configuration model that connects to out-of-tree providers via gRPC/TLS.
+- Make referent authentication explicit (e.g., authentication scope for cluster-scoped resources).
+- Add unified behaviors: refresh intervals, controller classes, retry settings, and gating policies.
+- Maintain ExternalSecret/PushSecret compatibility via apiVersion on storeRef.
+- Provide a migration path from v1 to v2, including a v1 plugin provider bridge and dedicated builds without v1 code.
+- Deliver all providers (e.g., AWS/GCP/Vault) as out-of-tree projects.
+- Migrate the existing end-to-end tests into the new out-of-tree structure
+- Provide documentation to validate and explain the new model.
+
+## Proposal
+
+### Overview
+
+ESO will run without bundled providers. Users deploy desired providers independently as separate services. ESO connects to providers over the network using secure gRPC/TLS.
+
+![Out of Tree](../assets/eso-out-of-tree.png)
+
+### API Resources
+
+#### SecretStore
+
+Remove `spec.provider` and introduce `spec.providerConfig`, which contains the endpoint and authentication required to reach an out-of-tree provider, plus a provider-owned reference forwarded on requests.
```

Comment:
I think the proposal is about not removing spec.Provider at first during migration period right?